### ISSUE TITLE
Feature Request: Delta Pro Ultra support 

### ISSUE THREAD

**Issue by RyanWor** [tolwi/hassio-ecoflow-cloud#193]
I would love to integrate my DPU with HA. I am happy to help with any discovery or testing.

**Comment by jasonjan2003** [comment]
I was able to set up node-red with my DPU, but everything just shows up as `Invalid JSON string`. I haven't used node-red before, but the nodes do show as connected, and entries in the debug log coincide with me manipulating the Ecoflow App. Any ideas?
```
2/9/2024, 12:19:13 PM[node: d9ecc011d0be04ea](https://homeassistant.address/api/hassio_ingress/.../#)
msg : string[19]
"Invalid JSON string"
```

**Comment by foxthefox** [comment]
Hmm, not so clear for me. The debug says something in relation data going into HA. If that is true then its far beyond the scope of test setup. But without seeing the node sketch, I can only guess.

Anyhow, as you wrote that the debug log coincide with the action via the App, its a good sign. So we have to focus on the data which is triggering the log.
If you have used the debug-nodes, then each of them posts its message into the right hand side log. The entry of that log starts with the topic and is followed by the message itself. May be it is necessary to click on the msg text or even a second time when there is only "object" written. After clicking and expanding the message, it should be readable text containing also {},[],:
If this is shown, we are very close to our goal. If you hover on the message, there are two small icons on the right edge. The one which is "Copy value" should be clicked and and the content pasted here.
There might be different JSON-strings and there might be some with the same structure/content. It would be great if we could catch one of each type.

**Comment by foxthefox** [comment]
Very good job. That is very useful and organized.
Unfortunately it is not userfriendly as the the nice JSON replies of the other devices.
But there is already some good starting point with the Powerstream, which is having similiar format.
The messages in HEX could be decoded with
https://protobuf-decoder.netlify.app
and in lucky case at least the header structure could be same with powerstream.
It looks in the first moment like that, but needs a deeper dive. First field is always the payload (the command) when it is from ..../set and at least that looks promising.

**Comment by jasonjan2003** [comment]
Cool! I looked into the `set_reply` messages with 737 elements, and here's what I identified:

Table 1
| Byte Range | Field Number | Type   | Content         | Name                                              |
| ---------- | ------------ | ---------- | --------------- | ------------------------------------------------- |
| 0.-87       | 1                 | protobuf |  <See Table 2> |                                                         |

Table 2
| Byte Range | Field Number | Type   | Content         | Name                                              |
| ---------- | ------------ | ---------- | --------------- | ------------------------------------------------- |
| 0.-73       | 1                 | protobuf |  <See Table 3> |                                                         |
| 73-75      | 1            | varint | As uint: 3      | Unknown                                           |
| 75-77      | 2            | varint | As uint: 0      | Unknown                                           |
| 77-79      | 1            | varint | As uint: 3      | Unknown                                           |
| 79-85      | 2            | varint | As uint: 0      | Unknown                                           |


Table 3: Mostly Settings
| Byte Range | Field Number | Type   | Content         | Name                                              |
| ---------- | ------------ | ------ | --------------- | ------------------------------------------------- |
| 0-2        | 1            | varint | As uint: 3      | Unknown                                           |
| 2-4        | 2            | varint | As uint: 0      | Unknown                                           |
| 4-6        | 3            | varint | As uint: 100    | Unknown                                           |
| 6-8        | 4            | varint | As uint: 0      | Backup reserve (0: Disabled, 1: Enabled)          |
| 8-10       | 5            | varint | As uint: 30     | Backup reserve level [%]                          |
| 10-12      | 6            | varint | As uint: 0      | Unknown                                           |
| 12-14      | 7            | varint | As uint: 60     | Backup reserve solar charging level [%]           |
| 14-16      | 8            | varint | As uint: 0      | Battery Preconditioning (0: Disabled, 1: Enabled) |
| 16-18      | 9            | varint | As uint: 90     | Charge Limit [%]                                  |
| 18-20      | 10           | varint | As uint: 10     | Discharge Limit [%]                               |
| 20-22      | 11           | varint | As uint: 1      | Unknown                                           |
| 22-24      | 12           | varint | As uint: 10     | Unknown                                           |
| 24-27      | 13           | varint | As uint: 1900   | AC Charging Speed [W]                             |
| 27-30      | 14           | varint | As uint: 900    | Power In/Out [W]                                  |
| 30-32      | 15           | varint | As uint: 120    | Device Timeout [min] (0=Never)                    |
| 32-36      | 16           | varint | As uint: 300    | LCD screen timeout [sec] (0=Never)                |
| 36-40      | 17           | varint | As uint: 240    | 12V DC Timeout [min] (0=Never)                    |
| 40-43      | 18           | varint | As uint: 0      | AC Timeout [min] (0=Never)                        |
| 43-46      | 19           | varint | As uint: 0      | Unknown                                           |
| 46-49      | 20           | varint | As uint: 1      | Unknown                                           |
| 49-53      | 21           | varint | As uint: 1199   | Unknown                                           |
| 53-71      | 22           | string | America/Chicago | Timezone                                          |

**Comment by foxthefox** [comment]
The structure of the message of powerstream is like that: 

`
message Header
{
    optional bytes pdata = 1;
    optional int32 src = 2;
    optional int32 dest = 3;
    optional int32 d_src= 4;
    optional int32 d_dest = 5;
    optional int32 enc_type = 6;
    optional int32 check_type = 7;
    optional int32 cmd_func = 8;
    optional int32 cmd_id = 9;
    optional int32 data_len = 10;
    optional int32 need_ack = 11;
    optional int32 is_ack = 12;
    optional int32 seq = 14;
    optional int32 product_id = 15;
    optional int32 version = 16;
    optional int32 payload_ver = 17;
    optional int32 time_snap = 18;
    optional int32 is_rw_cmd = 19;
    optional int32 is_queue = 20;
    optional int32 ack_type= 21;
    optional string code = 22;
    optional string from = 23;
    optional string module_sn = 24;
    optional string device_sn = 25;
}
`
Pdata the first slot, the payload. On issue #54 is some more information to it.

get_reply is mostly the answer from the device with status data and energy values. The commands are in set and set_reply is the ack to it. 

my next post will have some comments to yours, which was in the meantime showing up (while writing this on a train ride)

**Comment by foxthefox** [comment]
Comment to table 1,2,3
You perfectly got my idea. Table2 with field number 1 makes the payload, which is table 3. 
Sometimes one telegram can contain multiple messages. The 87 out 737 is the first message.
And luckily you already have some good hits for value naming👍

**Comment by foxthefox** [comment]
I have spent some time on the data.
The commands are quite clear now, based on the examples:
- DC_OUTPUT_USB -> cmdId 68
- AutoTimeout_Device -> cmdId 81
- AutoTimeout_LCDScreen -> cmdId 82
- AutoTimeout_AC -> cmdId 83
- AutoTimeout_12VDC -> cmdId 84
- ACChargingSpeed_AC and ACChargingSpeed_POWER -> cmdId 73
- EnergyManagement_HiLim -> cmdId 87
- EnergyManagement_LoLim -> cmd Id 88
- EnergyManagement_BACKUP -> cmdId 94 (inludes the low val and enable)
- 4G -> cmdId 117
- ACAlwaysOn -> cmdId 93 (includes a value of 15??, and enable)
- BatteryPreCond ->cmdId 89
- LabFeature -> cmdId 103
- Timezone(Town and Zone) -> cmdId86 for townId and 87 for the string

Interesting is the fact that the device serial number is not contained in the commands, that is different to the plug and powerstream.

What is not clear is the value 15 in ACAlwaysOn, maybe some percentage?

Above Table3 (message with cmdId=3 from get_reply) should be probably repeated with different modes of operation and settings to validate the data. I would guess that field 3, 12 is some kind of percentage, and field 6,11,19,20 could be some kind of on/off status. Field 1 could be some kind of error/status indicator and field 21 is the townId.

I have also checked the other parts of the get_reply.
My first guesses:
message with cmdId = 4
- field 5 capacity of battery
- field 8 charge limit
- field 9 discharge limit

message with cmdId = 2
- field 23 remaining capacity
- field 43 frequency
- field 61 battery voltage
- field 62 battery current
- field 75/77/79/81 ac voltages
- field 76/78/80/82 ac current

message with cmdId = 1
- field 24 power setting
- field 25 maximum power of setup
- field 41/42/56 temperatures if degF
- field 48/49/50 some kind of percenetages

The message with cmdId=28 is empty.

I am not sure what date the App provides, but hopefully some values of these messages are matching to the values shown in the App.

**Comment by jasonjan2003** [comment]
> What is not clear is the value 15 in ACAlwaysOn, maybe some percentage?

Indeed. After trying several values, this `15` turns out to be the minimum backup reserve percentage: 
<img src="https://github.com/tolwi/hassio-ecoflow-cloud/assets/5902451/4bb6644a-d19f-43f0-9103-26661b9158a3" alt="batteryReserve" width="300"/>
The minimum battery reserve value is set by the app to be 5% higher than the discharge limit (see image below). Since the app enforces this relationship on its own, it might be a good idea to perform this check for this HA integration.
<img src="https://github.com/tolwi/hassio-ecoflow-cloud/assets/5902451/397a74f9-7e30-4ad5-ae2e-f811539b4ec8" alt="energyManagement" width="300"/>

>message with cmdId = 4
>
(as float, very likely; someone with 2+ batteries can verify) field 5 capacity of battery
(correct) field 8 charge limit
(correct) field 9 discharge limit

>message with cmdId = 1
>

field 24 power setting, max ac charging power (uint) 
field 25 maximum power of setup, seems correct, max output of setup (x1 inverter + x1 battery), using "Power In/out" port
field 41/42/56 ~temperatures if degF~ 41: total input power [W], float (but rounded to integer?); 42: total output power [W], float; 56: total input power [W], float
field 48/49/50/51 ~some kind of percenetages~, AC 20A output power [W], float, from left to right looking at the inverter LCD. I would guess 52/53/54 are the AC 20A/AC30A/POWER IO output power [W] values. I don't have the plugs to test them right now.

A new function command for renaming the unit (`cmdId=65`)
| Function | Action | Debugnode# | Topic                                                | PayloadSize | Payload                                                                                                    | QOS | Retain |
| -------- | ------ | ---------- | ---------------------------------------------------- | ----------- | ---------------------------------------------------------------------------------------------------------- | --- | ------ |
| Rename   | "HomeBattery" | 5          | /app/MQTT_USER_ID/DEVICE_ID/thing/property/set       | 53          | 0A330A0D0A0B486F6D65426174746572791020180220012801380340024841500D580170BCBD93D001800113880101BA0103696F73 | 0   | FALSE  |
| Rename   | "HomeBattery" | 6          | /app/MQTT_USER_ID/DEVICE_ID/thing/property/set_reply | 35          | 0A2110021820200128013803400248415801600170BCBD93D00178813A800103880101                                     | 0   | FALSE  |

**Comment by foxthefox** [comment]
I think we are good progressing.
Can you check if the 5% margin for minimum battery reserve value is automatically adjusted when you change the discharge limit? Meaning that in the condition of switched ON of Backup Reserve there will be 2 outgoing set-telegrams, one for discharge level and one for the backupreserve.

I think the float 56 is not the total input power, it is more the AC-In power, which is the total when no other source is connected.

I think we should also check some USB charging resulting in power, volt, amps.
Based on your feedback I have made a new guess and implementation.
When using the last get_reply from the csv I could find such a matching

```
{
  dpu_3: {
    unknown1: 3,
    unknown2: 0,
    unknown3: 100,
    bkpResEn: 0,
    bkpResLev: 30,
    unknown6: 0,
    bkpResSolLev: 60,
    batPreCondEn: 1,
    chaLim: 90,
    dischaLim: 10,
    unknown11: 1,
    unknown12: 10,
    acSpeed: 1800,
    power: 900,
    devTimeout: 120,
    lcdTimeout: 300,
    dc12vTimeout: 720,
    acTimeout: 0,
    unknown19: 0,
    unknown20: 1,
    town: 1199,
    area: 'America/Chicago'
  },
  dpu_4: {
    unknown1: 1,
    unknown3: 89,
    unknown4: -27,
    capacity5: 6144,
    unknown7: 21398,
    chaLim: 90,
    dischaLim: 10,
    unknown11: 28
  },
  dpu_1: {
    sysVer1: 16777219,
    unknown16: '',
    unknown21: 89,
    maxAcPower24: 1800,
    maxPowerPowerPort25: 7200,
    totalInputPower41: 168,
    totalOutputPower42: 146,
    usbC1power43: 0,
    usbC2power44: 0,
    usbA1power45: 0,
    usbA2power46: 0,
    dcPower47: 0,
    ac1outBackupPower48: 67.21940612792969,
    ac2outBackupPower49: 41.767330169677734,
    ac1outOnlinePower50: 36.865638732910156,
    ac2outOnlinePower51: 0,
    acout30APower52: 0,
    acoutL14Power53: 0,
    ioPower54: 0,
    totalInputPower56: 167.6139373779297,
    solarLowInputPower57: 0,
    solarHighInputPower58: 0,
    float59: 0
  },
  dpu_2: {
    unknown1: 0,
    unknown21: 4,
    watthBatt23: 2148,
    freq43: 60,
    watthOut46: 105132,
    watthIn47: 108132,
    freq48: 60,
    volt61: 106.62699890136719,
    current62: -0.2619999945163727,
    float63: 0,
    float64: 27,
    usbC1Volt65: 0.8860361576080322,
    usbC1Current66: 0.0001125011476688087,
    usbC2Colt67: 0.8860361576080322,
    usbC2Current68: 0.010011250153183937,
    usbA1Volt69: 0.018721750006079674,
    usbA1Current70: 1.800022533370721e-13,
    usbA2Volt71: 0.025763917714357376,
    usbA2Current72: 0,
    dcVolt73: 1.5700000524520874,
    dcCurrent74: 0,
    ac1outBackupVolt75: 120.96668243408203,
    ac1outBackupCurrent76: 0.5222846269607544,
    ac2outBackupvolt77: 120.96668243408203,
    ac2outBackupCurrent78: 0.3502686023712158,
    ac1outOnlineVolt79: 120.6103744506836,
    ac1outOnlineCurrent80: 0.2961742877960205,
    ac2outOnlineVolt81: 120.6103744506836,
    ac2outOnlineCurrent82: 0,
    ac30AVolt83: 120.96668243408203,
    ac30ACurrent84: 0,
    acL14Volt85: 241.57705688476562,
    acL14Current86: 0,
    solarLowVolt87: 0,
    solarLowcurrent88: 0,
    solarHighVolt89: 0,
    solarHighCurrent90: 0,
    acInVolt91: 121.90068054199219,
    acInCurrent92: 1.3557208776474,
    powerIOInVolt93: 0,
    powerIOInCurrent94: 0,
    powerIOOutVolt95: 0,
    powerIOOutCurrent96: 0,
    temp97: 41,
    temp98: 42,
    temp99: 42,
    temp100: 41
  }
```
The number at the end is always the field number.
Temperatures could also be in °C, if that would fit.

You can also post another get_reply message for me to have another guess.

**Comment by foxthefox** [comment]
One more question, the USB-C and USB-A can only be switched via the HW-button on the device, but is there an indication to the off/on status in the App? So we have to search for it.

**Comment by muenchris** [comment]
Has anyone decoded the heartbeat telegrams yet that come in on /app/devices/property/<SerialNumber> ?
I see four command IDs:
1, 28, 2 and 4.
But the pData field contains no protobuf decoded message like the other telegrams coming in from get_reply.
There is something in the pData field but cannot be decoded with the Protobuf Decoder.

**Comment by muenchris** [comment]
message with cmdId = 4

field 3 is SoC (state of charge)
field 7 is minutes remaining
(as float, very likely; someone with 2+ batteries can verify) field 5 capacity of battery
(correct) field 8 charge limit
(correct) field 9 discharge limit

**Comment by foxthefox** [comment]
> Has anyone decoded the heartbeat telegrams yet that come in on /app/devices/property/SerialNumber ? I see four command IDs: 1, 28, 2 and 4. But the pData field contains no protobuf decoded message like the other telegrams coming in from get_reply. There is something in the pData field but cannot be decoded with the Protobuf Decoder.

Nice to see that you are able to see the telegrams and that you can crosscheck/validate assumptions.
In general there are cmdId1,2,3,4 where we made already good progress, cmdId 28 is somehow empty.
What we know so far, you can see in the previous comments.
As I see, SoC and remainTime is demystified additionally.
For evaluation I made some small programs, if you could post some telegrams and the changed condition would be great.
E.g.
USB-C turned on -> get_reply message
USB-C turned off -> get_reply message
USB-A-turned on  -> get_reply message
USB-A turned off -> get_reply message
DC turned on -> get_reply message
DC turned off -> get_reply message
Solar turned on -> get_reply message (and hopefully solar connected)
.....
With that the status should be shown somewhere and the voltage/current/power should show up.
So we would progress in demystifing the telegrams.

**Comment by muenchris** [comment]
Yes I can decode some of the values in cmd 1-4 from your descriptions above and my own research.
But I do have to send a "get" publish to the MQTT server in order to get valid 1-4 telegrams.
There are hearbeat telegrams coming in on the /app/devices/property/serial that the app seams to use to update the screen with current AC and DC in/out power values. But the pdata is not in the same format as the 1-4 telegrams when responding to the "get" message.

I ran the pdata through the https://protobuf-decoder.netlify.app/ decoder but it could not make sense of it.
cmdid 28 is empty during the hearbeat as well. but 1,2 and 4 contain some data in pdata - might either be encrypted or not protobuf:

CMDID2: sends:
ab a3 0b a2 a6 13 a2 b3 1b a2 43 b3 63 a2 a3 6b a2 a2 6b a1 a3 73 a1 a3 7b a1 a3 43 a1 a2 4b a1 a2 53 a1 1a 13 a5 5b a1 52 64 a5 23 a0 9f 2b a0 a3 33 a0 a3 3b a0 a3 03 a0 a3 0b a0 a3 4e a0 6c 54 70 e1 56 a0 9e a9 d4 1c 5e a0 a3 a3 a3 a3 26 a7 a3 a3 6f e1 2e a7 2e 3a c4 99 36 a7 19 c1 cd 98 3e a7 2e 3a c4 99 06 a7 c8 23 ce 98 0e a7 3f 9a 97 8a 16 a7 34 6b 9f 9a 1e a7 fb 9a 97 b5 66 a7 a3 a3 a3 a3 6e a7 a3 a3 a3 a3 76 a7 a3 a3 a3 a3 7e a7 b7 7f 4d e1 46 a7 a3 a3 a3 a3 4e a7 b7 7f 4d e1 56 a7 a3 a3 a3 a3 5e a7 8d de 51 e1 26 a6 a3 a3 a3 a3 2e a6 8d de 51 e1 36 a6 a3 a3 a3 a3 3e a6 b7 7f 4d e1 06 a6 29 1f 50 9c 0e a6 02 0f d3 e0 16 a6 a3 a3 a3 a3 1e a6 a3 a3 a3 a3 66 a6 a3 a3 a3 a3 6e a6 a3 a3 a3 a3 76 a6 a3 a3 a3 a3 7e a6 c6 bc 52 e1 46 a6 29 13 98 9e 4e a6 a3 a3 a3 a3 56 a6 a3 a3 a3 a3 5e a6 a3 a3 a3 a3 26 a5 a3 a3 a3 a3 2e a5 a3 a3 ab e1 36 a5 a3 a3 ab e1 3e a5 a3 a3 ab e1 06 a5 a3 a3 ab e1 0b a5 ba 16 a5 a3 a3 a3 a3

CMDID1: much shorter but much more frequent sends:
d3 04 06 06 64 45 a3 05 9c e6 67 45

CMDID4: sends
9d 80 9f 96 8f da b2 97 97 31 55 ba 97 97 57 d2 af 2d 8b d7 f3 df 90 cf 9b

**Comment by foxthefox** [comment]
These extractions are somehow not decodedable (maybe some missing starting bytes). Could you post the whole get_reply messages which are shorter but frequently incomming?

For sending the request for whole update, I will provide some advice.

**Comment by foxthefox** [comment]
The following approach should work for the request of full update telegrams.
The ef-mqtt should be copied (it stays with the broker configuration node) and an inject node must be connected to the newly copied publishing node.
It should look like this.
![Bildschirmfoto 2024-02-16 um 08 14 39](https://github.com/tolwi/hassio-ecoflow-cloud/assets/16841643/b8befcf9-8609-4e66-bd8b-65ea8837b99e)

Inside the inject node
- the msg.topic should be the /app/mqttuserId/serial#/get
- the msg.payload must be changed to buffer
- the msg.payload should contain [10,16,16,32,24,32,112,132,190,167,136,1,186,1,3,105,111,115] for making the request.

![Bildschirmfoto 2024-02-16 um 08 15 11](https://github.com/tolwi/hassio-ecoflow-cloud/assets/16841643/9af6f0d2-9990-4466-8b6f-70d064aae5fc)

The get request contains a number (... ,132,190,167,..) which is an identifier for the message and could be found in the reply. It might be necessary to change the number for each request, but I would think that it is not needed (only the presence of it is needed).

**Comment by muenchris** [comment]
Sorry but I am a CSharp developer integrating this into another project and am not familiar with NodeRed.

All the forced "get" messages that reply with "get_reply" can be encoded just fine and you have done all these already.

What's missing are the hearbeat telegrams that come in on /app/device/property/serial# and do not require any publish on any kind.

Right now my workaround is to publish the "get" command periodically and then get the "get_reply" back. That works great but I rather be a "silent" listener on the heartbeat.

**Comment by muenchris** [comment]
Here the full HEX from the hearbeats:
CmdID: 1
0a 2e 0a 0c 21 f6 f4 74 68 b7 51 f7 b4 60 ab b7 10 02 18 20 20 01 30 01 38 03 40 02 48 01 50 0c 58 01 70 f4 d3 f6 01 78 81 3a 80 01 03 88 01 01 

This CMDID: 2
0a da 02 0a b6 02 ee e6 4e e7 e3 56 e7 f6 5e e7 02 f6 26 e7 e6 2e e7 e7 2e e4 e4 36 e4 e6 3e e4 da 06 e4 e7 0e e4 e7 16 e4 2d 5e e0 1e e4 65 36 e0 66 e5 da 6e e5 e6 76 e5 e6 7e e5 e6 46 e5 e6 4e e5 e6 0b e5 1e f5 30 a4 13 e5 f4 65 26 58 1b e5 e6 e6 e6 e6 63 e2 e6 e6 c6 a4 6b e2 e6 f2 f6 da 73 e2 9a 26 93 dd 7b e2 e6 f2 f6 da 43 e2 01 1d cf da 4b e2 36 47 2d dd 53 e2 28 4e 71 d1 5b e2 43 88 3a dd 23 e2 e6 e6 e6 e6 2b e2 e6 e6 e6 e6 33 e2 e6 e6 e6 e6 3b e2 fd 67 0b a4 03 e2 e6 e6 e6 e6 0b e2 fd 67 0b a4 13 e2 ae f6 d9 d9 1b e2 73 6c 14 a4 63 e3 e6 e6 e6 e6 6b e3 73 6c 14 a4 73 e3 e6 e6 e6 e6 7b e3 fd 67 0b a4 43 e3 c2 40 16 d9 4b e3 3e e3 96 a5 53 e3 e6 e6 e6 e6 5b e3 e6 e6 e6 e6 23 e3 e6 e6 e6 e6 2b e3 e6 e6 e6 e6 33 e3 e6 e6 e6 e6 3b e3 75 d9 09 a4 03 e3 1d dc 34 d9 0b e3 e6 e6 e6 e6 13 e3 e6 e6 e6 e6 1b e3 e6 e6 e6 e6 63 e0 e6 e6 e6 e6 6b e0 e6 e6 de a4 73 e0 e6 e6 ce a4 7b e0 e6 e6 ca a4 43 e0 e6 e6 d6 a4 4e e0 c4 53 e0 e6 e6 e6 e6 10 02 18 20 20 01 28 01 30 01 38 03 40 02 48 02 50 b6 02 70 e6 d5 f6 01 78 81 3a 80 01 03 88 01 01 

And CMDid=4:
0a 3b 0a 19 ed f0 ef e6 ff bf c2 e7 e7 eb 24 ca e7 e7 27 a2 df 70 c7 a7 b9 af e0 bf f7 10 02 18 20 20 01 28 01 30 01 38 03 40 02 48 04 50 19 70 e7 d5 f6 01 78 81 3a 80 01 03 88 01 01

**Comment by foxthefox** [comment]
However, if you are able to get the get_reply messages then let’s still try to demystify all the values which are incoming. The above is partly a follow up of the idea of the left 4 sockets (as per our last messages) and is not validated. Despite having said the above, this interpretation only covers approximately half of the possible values. What’s most important are the positions of the values which have commands (needed for initialising).

I agree that it would be good to listen only to the frequent incoming updates which is the follow up of the initialising with the get-reply message.

However, I do not own the DPU, so I am only able to participate when others that do contribute messages for evaluation.

So it would be helpful if you could sent examples of the frequent incoming messages and get_replies after changing one condition at a time ( see above exemplary sequence).

**Comment by foxthefox** [comment]
What could be seen in the heartbeats:
- usage/appearance of "encType":1,
- "version":3, usually it is 19 when comming from the App
Especially encType supports the assumption that the pdata is encrypted.

**Comment by muenchris** [comment]
Yes It looks like the messages are encrypted in the pdata. The app must be able to read them as it updates the main screen without sending the "get" messages. Only on screen change the app sends the "get". This is generally a good design as it keep the amount of incoming messages to the MQTT broker small. The encryption is annoying but generally also a good measure from Ecroflow. Their public API - for which you need a developer account - does not seam to have the a heartbeat and uses JSON in all MQTT telegrams, but I am still waiting for my approval from Ecoflow to verify.

**Comment by jasonjan2003** [comment]
> I think we are good progressing. Can you check if the 5% margin for minimum battery reserve value is automatically adjusted when you change the discharge limit? Meaning that in the condition of switched ON of Backup Reserve there will be 2 outgoing set-telegrams, one for discharge level and one for the backupreserve.

When `Backup reserve` is enabled:
1. if `Discharge limit <= (Backup reserve - 5%)`, only 1 **set** message is sent:
   - cmd Id 88 to new `Discharge limit` value.
2. if `(Backup reserve - 5%) < Discharge limit`, 2 **set** messages are sent:
   - cmd Id 94 to 
      - enable (1), and
      - new `Backup reserve` value, then
   - cmd Id 88 to new `Discharge limit` value.

>I think the float 56 is not the total input power, it is more the AC-In power, which is the total when no other source is connected.

That makes a lot of sense! I don't have another input source right now.

Switching between `DC Output/USB` switch on the app:
| dc output / usb disabled | 0A550A47080310001864200028233000383C4001485A500A5801600A68EC0E70840778788001AC028801F001900100980100A00101A801AF09B2010F416D65726963612F4368696361676F10024002480370B3FA81D2030A280A1A0A180801185A250000B8C12D0000C04538A2A801405A480A581C10024002480470B3FA81D2030A0C10064003481C70B3FA81D2030A87020AF8010883808008109412180450005800600068007001780A820100A8015AB00101B80101C001880EC801A038D001A2A801D80100E00164E8011ECD0200002543D50200001243DD0200000000E50200000000ED0200000000F50200000000FD020000000085036E8A87428D03A86222429503A10B14429D0300000000A50300000000AD0300000000B50300000000BD0300000000C503FA072543CD0300000000D50300000000DD0300000000A00602A80600B00600B806FFFFFFFF0FC006FFFFFFFF0FC806FFFFFFFF0FD00600D80600E00600E80600C80700D007FFFFFFFF0FD807FFFFFFFF0FE007FFFFFFFF0FE80700F00700F8070080080010024002480170B3FA81D2030AC5020AB6020800A80104B00110B801E410C00100C80101C80202D00200D8023CE00201E80201F00289B806F802C1CF0680033C880300900300980300A00300A80300ED0339F4D542F503F85363BEFD030000000085040000B8418D046986B13C95043CC33C379D046986B13CA50483D7233CAD04F60BA03CB5043298F135BD04AE0C433CC50400000000CD04EC51B83ED50400000000DD04505AF242E504D298053FED04505AF242F5042A77AA3EFD04AB9AF04285059384993E8D05AB9AF0429505000000009D05505AF242A50500000000AD057E7A7143B50500000000BD0500000000C50500000000CD0500000000D50500000000DD05AD3CF442E505589BAA3FED0500000000F50500000000FD05000000008506000000008D0600003042950600002C429D0600003442A50600003042A80627B5060000000010024002480270B3FA81D203 |
| ------------------ | ---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- |
| dc output / usb enabled     | 0A550A47080310001864200028233000383C4001485A500A5801600A68EC0E70840778788001AC028801F001900100980100A00101A801AF09B2010F416D65726963612F4368696361676F10024002480370EBAA87D2030A280A1A0A180801185A2500003CC22D0000C045389DA801405A480A581C10024002480470EBAA87D2030A0C10064003481C70EBAA87D2030A87020AF8010883808008109612180450005800600068007001780A820100A8015AB00101B80101C001880EC801A038D0019DA801D80100E00164E8011ECD0200002743D50200002243DD0200000000E50200000000ED0200000000F50200000000FD02000000008503FD84AC428D0393FF20429503345E11429D0300000000A50300000000AD0300000000B50300000000BD0300000000C50327322743CD0300000000D50300000000DD0300000000A00602A80600B00600B806FFFFFFFF0FC006FFFFFFFF0FC806FFFFFFFF0FD00600D80600E00600E80600C80700D007FFFFFFFF0FD807FFFFFFFF0FE007FFFFFFFF0FE80700F00700F8070080080010024002480170EBAA87D2030AC5020AB602081FA80104B00110B801E410C00100C80101C80202D00200D8023CE00201E80201F002F7B706F802AFCF0680033C880300900300980300A00300A80300ED03DDE4D542F5030AD743BFFD0300000000850400003C428D04EB70A24095047B549E2B9D04EB70A240A504E8CB4C3CAD04548AA240B504BC9BEE3BBD04FDAEA040C504A6CF233CCD0400004C41D50400000000DD048109F342E5042C67143FED048109F342F5049DE2A93EFD049B5AF1428505AAF3A03E8D059B5AF1429505000000009D058109F342A50500000000AD050E327243B50500000000BD0500000000C50500000000CD0500000000D50500000000DD0542B0F542E505B677B03FED0500000000F50500000000FD05000000008506000000008D0600003042950600002C429D0600003442A50600003042A80627B5060000000010024002480270EBAA87D203 |

Some observations:
- `dpu_4`, `remainTime7` in minutes as uint. I only have 1 battery, but it tracks exactly. This value can be either time until "fully" charged, or time until battery is "fully" discharged. (Fully, as defined by user-defined limits)
-  `dpu_1`, `remainTime26` seems to be copy of `dpu_4`, `remainTime7` (above).
- `dpu_4`, `isCharging2` **might be** whether the battery is being charged, as boolean uint. This field is not shown when the battery is not being charged. Some one with a different input source should validate.
- `dpu_1`, `unknown1` turned to uint 31 when `dc output /usb` was enabled via the app (see table above). Otherwise, it was 0.
- Thanks everyone. We're making good progress!

**Comment by michaelahern** [comment]
Was digging into the heartbeats and pdata values as well. It looks like the decoding mechanism is to XOR the pdata bytes with the seq field from the message header. Something as simple as:

```
header.msg.pdata.forEach((element, index) => {
    pdataDec[index] = element ^ header.msg.seq;
});
```

You should be able to run the output through the protobuf decoder afterwards. Attaching protobuf message defs for the following heartbeat cmdId's...

1: [AppShowHeartbeatReport.proto](https://github.com/tolwi/hassio-ecoflow-cloud/files/14321255/AppShowHeartbeatReport.proto.txt)
2: [BackendRecordHeartbeatReport.proto](https://github.com/tolwi/hassio-ecoflow-cloud/files/14321256/BackendRecordHeartbeatReport.proto.txt)
3: [AppParaHeartbeatReport.proto](https://github.com/tolwi/hassio-ecoflow-cloud/files/14321254/AppParaHeartbeatReport.proto.txt)
4: [BpInfoReport.proto](https://github.com/tolwi/hassio-ecoflow-cloud/files/14321257/BpInfoReport.proto.txt)

**Comment by michaelahern** [comment]
Also, for reference, attaching the [protobuf file descriptor](https://github.com/tolwi/hassio-ecoflow-cloud/files/14321318/pbdesc.txt) used to regenerate the above messages. Run this through the protobuf decoder as well.

**Comment by muenchris** [comment]
> Was digging into the heartbeats and pdata values as well. It looks like the decoding mechanism is to XOR the pdata bytes with the seq field from the message header. Something as simple as:
> 
> ```
> header.msg.pdata.forEach((element, index) => {
>     pdataDec[index] = element ^ header.msg.seq;
> });
> ```
> 
> You should be able to run the output through the protobuf decoder afterwards. Attaching protobuf message defs for the following heartbeat cmdId's...
> 
> 1: [AppShowHeartbeatReport.proto](https://github.com/tolwi/hassio-ecoflow-cloud/files/14321255/AppShowHeartbeatReport.proto.txt) 2: [BackendRecordHeartbeatReport.proto](https://github.com/tolwi/hassio-ecoflow-cloud/files/14321256/BackendRecordHeartbeatReport.proto.txt) 3: [AppParaHeartbeatReport.proto](https://github.com/tolwi/hassio-ecoflow-cloud/files/14321254/AppParaHeartbeatReport.proto.txt) 4: [BpInfoReport.proto](https://github.com/tolwi/hassio-ecoflow-cloud/files/14321257/BpInfoReport.proto.txt)

Nice! I was hoping they do a simple XOR as that is the simplest version of "obfuscation". I will give it a try tomorrow with my code.

Thank you!

**Comment by foxthefox** [comment]
@jasonjan2003 
dc_on changed also the dcVolt73 or outAdsVol (new protobuf) to 12.75, which proves the DC output.
the unknown1 (changed to 31) is now called recordFlag so it might not be the right indicator for dc is on.
but I discovered a change in showFlag from 2324 to 2326 so the 2nd bit could be the indicator for DC on/off

@michaelahern 
Awesome proto files and the XOR works.
Maybe some adjustmenst needed, the 

in cmdId=1 or now AppShowHeartbeatReport, the xxTimeTaskType, xxTimeTaskIndex, xxTimeTaskMode is to be treated as unint32, which gives a strange number.
I have changed it to int32 at the chg.. and it returns a "nice" -1, most likely the indication that there is no task setup.
So this might be crosschecked with a device with time task setup. 

Here is the output:
```
    chgTimeTaskType: -1,
    chgTimeTaskIndex: -1,
    chgTimeTaskMode: -1,
    chgTimeTaskParam: 0,
    chgTimeTaskTable_0: 0,
    chgTimeTaskTable_1: 0,
    chgTimeTaskTable_2: 0,
    dsgTimeTaskNotice: 0,
    dsgTimeTaskType: 4294967295,
    dsgTimeTaskIndex: 4294967295,
    dsgTimeTaskMode: 4294967295,
    dsgTimeTaskParam: 0,
    dsgTimeTaskTable_0: 0,
    dsgTimeTaskTable_1: 0,
    dsgTimeTaskTable_2: 0
```

**Comment by foxthefox** [comment]
While implementing the data model, I got some questions.

What is the response when the DPU is not connected to the Cloud (no Internet access, powered off)? When checking the "get" for a disconnected powerstream gives a short message and when decoded it is -2.

There is a "solarOnlyFlag" in the dataset, can this mode be activated from the App?
There is a "sysWorkMode" in the dataset, can the work mode changed via App?

Is there an AC Discharge direct activation from the App? Looking at the time task it seems that there is also a AC discharge mode can be scheduled.

**Comment by muenchris** [comment]
Not sure was you mean by "AC Discharge" what would that do? There is a AC on/off button but that turns all AC Ports on/off. I would love to have that for each port individually but dont see that in the App.
The "Automation" tab allows to set a "AC Discharge "If then" time of day based rule. I assume it will turn AC on at the time set, but there is no "off" setting - not quite clear what that will do without the "off".
The same can be set for AC Charging and DC Discharging. 

Solar Only is kind of there: The app has an "Energy Management" tab. It allows to set a "Backup reserve". If that is set to 12% (which is the minimum as I can tell), then the DPU only uses Solar until its down to 12% below that number it takes power from the AC Side. If the Backup reserve is off, it always uses all available charging sources.

Can you elaborate what you understand as "Work Mode"? I dont see any setting with this name in the app, but maybe it has a different name.

As for the response if the DPU is not connected to the cloud, I need to do test this when I am back from my trip (in three weeks ). Maybe somebody else with the DPU can test this

**Comment by foxthefox** [comment]
Thank, you. Exactly in time tasks also called automation there is AC Discharge,Charge and DC Discharge
 and I thought that this kind of switching off/on can also be done elsewhere.

In "AppParaHeartbeatReport" is "sys_work_mode", I have renamed it from word to work.

**Comment by RyanWor** [comment]
Apologies that I have been fairly inactive since opening this issue. I have a second DPU now as well as a SHP2 (the SHP2 should be installed in the next week or two). I am happy to help test or troubleshoot anything if it will help develop support for these devices. Thanks!

**Comment by foxthefox** [comment]
I was working on DeltaProUltra in the meantime, when I am sitting in the train back home, I will make further explanations in next comment.

I also worked with SHP, so maybe there are some matching parts.
For the SHP2 I suggest to do similar investigation as the ones for DPU. I would guess that SHP2 is also using protobuf, but we will see during investigation.

**Comment by RyanWor** [comment]
> I was working on DeltaProUltra in the meantime, when I am sitting in the train back home, I will make further explanations in next comment.
> 
> I also worked with SHP, so maybe there are some matching parts. For the SHP2 I suggest to do similar investigation as the ones for DPU. I would guess that SHP2 is also using protobuf, but we will see during investigation.

My SHP2 should be installed in the next two weeks. Once it is up and running I will setup a test environment and see if I can contribute some investigation on the SHP2 values. Thanks again for all your work on this!

**Comment by foxthefox** [comment]
Is your SHP2 installed yet?

**Comment by RyanWor** [comment]
> Is your SHP2 installed yet?

The install was just finished late Saturday evening. I haven't had much time to play around with it yet given the holiday yesterday. I am also still waiting on one additional power I/O cable to come in middle of this week for my second DPU/battery but that may not matter much for testing purposes. 

I plan to review this thread and setup test/investigation environment this weekend so I can help contribute findings. If there is anything specific beyond what's already been covered here I need to do or specific points of information you would like back please let me know. I will try and get it done by next week. Thanks!

**Comment by RyanWor** [comment]
Sorry it's going to be a little bit longer before I can setup the test environment and assist on the SHP2 front. I got a serious unexpected real life test of the system as we got hit hard by sudden storms w/ 120+ mph wind in Houston late last week and were without power until early this week (the SHP2 and dual DPU/batteries worked wonderfully to cover the gaps in generator power). I am just now getting back to work and trying to return to normal life so just need a bit longer on my end to assist here. Apologies for the delays and thanks for the patience.

**Comment by foxthefox** [comment]
sad to hear about the storm, but glad that the EF ecosystem has leaveraged the situation. take your time we are not in a hurry

**Comment by foxthefox** [comment]
I think that you should change the message type from JSON to buffer in the mqqt-in node. Then you will receive strange strings in the debug log. But you also have a small icon „raw“ at the message. Click on this and it should be a HEX string which we need.

**Comment by RyanWor** [comment]
> I think that you should change the message type from JSON to buffer in the mqqt-in node. Then you will receive strange strings in the debug log. But you also have a small icon „raw“ at the message. Click on this and it should be a HEX string which we need.

Ah okay, thank you. That's what I just tested a bit ago and I am indeed getting the decimal strings now. If I expand a debug message and click "raw" 

I don't seem to be getting hex output tho, rather lots strange characters:
![image](https://github.com/tolwi/hassio-ecoflow-cloud/assets/757232/bb90cd98-c55e-4376-9ac4-93badb1a2ce9)

I think I actually need to leave it as string and then I can see the hex output, might just require copying them one by one to build the full string out for each payload. 
![image](https://github.com/tolwi/hassio-ecoflow-cloud/assets/757232/bcd6e95e-5939-4764-bdd6-2508ad2c22ca)

**Comment by RyanWor** [comment]
Okay I think I figured out a way to make that easier. Had GPT-4o create code for Excel VBA to do the conversion. Tested it with Jason's values from his excel first just to confirm it was working correctly (it had a formatting issue at first but was able to get it seemingly perfect in the next iteration. 

Here is the VBA code just in case anyone else finds it's useful in the future:
```
Function DecToHexString(decString As String) As String
    Dim decArray() As String
    Dim hexString As String
    Dim i As Integer
    Dim hexValue As String
    
    ' Remove brackets and split the string into an array
    decString = Replace(decString, "[", "")
    decString = Replace(decString, "]", "")
    decArray = Split(decString, ",")
    
    ' Convert each decimal to hex and build the hex string
    For i = LBound(decArray) To UBound(decArray)
        hexValue = WorksheetFunction.Dec2Hex(CLng(decArray(i)), 2)
        hexString = hexString & " " & hexValue
    Next i
    
    ' Trim leading space and return
    DecToHexString = Trim(hexString)
End Function

```

**Comment by foxthefox** [comment]
Good progress, I think if you get into the line called buffer, all the values are in one set of 233 values. Or it is  needed to click on rectangle “raw“.

**Comment by RyanWor** [comment]
Okay so starting to collect values on the SHP2 now. Looks like the main payload is pretty big (1896). When I paste it in hex format into the protobuf decoder it only seems to decode the first 129 values which equates to about 13 fields. The data I can decode looks good, pretty sure I already spotted grid input wattage in some of that. However, the rest of the payload is shown as "Left over bytes". 

I know there has to be more in this output since I can see things like my time zone, address, and GPS coordinates as plain text when I look at those payloads in the raw output.

**Comment by RyanWor** [comment]
Gonna go ahead and share what I have just in some initial testing so far but since I know this data contains some personal information of mine (address, GPS coordinates, etc.) I am going to put it in a Google Sheets and can grant your access once you click the link and request access: https://docs.google.com/spreadsheets/d/1RGt7lxu7dCPBS62eblS5QGBN6Zff6rr21oQwyhx8fKo/edit?usp=sharing

**Comment by foxthefox** [comment]
I will create tomorow a google account.

> Okay so starting to collect values on the SHP2 now. Looks like the main payload is pretty big (1896 length). When I paste it in hex format into the protobuf decoder it only seems to decode the first 129 values which equates to about 13 fields. The data I can decode looks good, pretty sure I already spotted grid input wattage in some of that. However, the rest of the payload is shown as "Left over bytes".
> 
> I know there has to be more in this output since I can see things like my time zone, address, and GPS coordinates as plain text when I look at those payloads in the raw output.

Usually there should be no "left over bytes". I would guess that there is something not correctly derived.
I have checked another way of convenient capturing. If the mqtt-in node is set to "base64 string" it gives still some unreadable string, but better than a normal string. This base64 encoded string can be copied when hoovering over the right icon "Copy text". This base64 text can also be pasted into the protobuf decoder. After let it decode, it also shows a nice HEX string. For comparison reasons you should try this way also, there should be no "left over bytes" with this approach.

**Comment by RyanWor** [comment]
Tested with base64 and pasting that into protobuff decoder and still get a lot of leftover bytes. Maybe node red or EcoFlow is truncating the payload and it spills over into another message? If you'd rather me share through Office/OneDrive I can do that too if it's easier for you and you don't have to create a Google account.

**Comment by RyanWor** [comment]
Just granted you access. Also pasted in the base64 encoded values I got yesterday after changing my node-red nodes/flow. I have to run here in a bit but if you want me to pull anything further once you have looked over the data I got so far I can do that when I get back later today. Thanks again for your help!

**Comment by foxthefox** [comment]
Thanks, I already started to decode and to put some data back to the table.

first insight:
- telegrams are similiar but different, assuming it uses the same header the cmdFunc and cmdId are always the same, that was at other devices used to disinguish between the commands
- the command data seems now to be given as field# and value instead of plain value and individual cmdFunc/Id
- deviceId is not transmitted in the header
- replies contain encoded data 

would be good to have some cyclic telegrams at debug2

More things I will directly put into the table

**Comment by foxthefox** [comment]
I saw that the chat within google docs is not persistent and in case you might missed the part for enlarging the debug-node size.
The default size of debug is 1000 and it must be changend in the settings.js to e.g. 2800 when using base64.
The file "settings.js" is normally to be found in the root of node-red -> .node-red/
The parameter "debugMaxLength" is almost at the end of the file.

https://nodered.org/docs/user-guide/runtime/configuration
https://nodered.org/docs/user-guide/runtime/settings-file

**Comment by RyanWor** [comment]
Yes, I pasted new complete values in the spreadsheet after changing max debug length to 5000. I have not had time to start digging into the values yet though. We had another storm and power outages yesterday and now my SHP2 is reporting one of the relays is faulty. So working with support to get that fixed/replaced. I should still be able to start working through figuring out values in the meantime but it may be next week before I can make time to go through them in earnest. Thanks again for all your help setting this up!

**Comment by RyanWor** [comment]
Had some extra time today after all. Got quite a lot of the table built out for Debug4. Still have lots of the heavily nested protobuffs in the last half of the page to add to the table. Started adding remarks for things I've identified so far. It will probably take quite a lot of time to decipher all of these but I will keep working away at it little by little as time allows. Thanks!

**Comment by foxthefox** [comment]
I will spend some time today for creating the protobuf structure according your very good investigation. This will create a JSON object with value name and setting. Based on this I have another script to compare telegrams for changes, which is handy for finding command values or other singular changes.

**Comment by RyanWor** [comment]
Oh yes if you have a script to detect changes that would probably so much easier than the manual way I have been grabbing telegrams and comparing side by side. Debug 4 is quite big on the SHP2 so it's a lot of scrolling and hunting around.

**Comment by foxthefox** [comment]
I am still creating the structure file. And yes, the 2nd message in debug4 is large and deeply nested. I will create a new sheet and paste some JSON strings into it for evaluation/guessing.

**Comment by foxthefox** [comment]
I have created a proto file matching the protobuf decoding fields and value types.
Unfortunately the SHP2 is not using similar structures as the older version, so many things are unknown and some can be assumed also being part of the telegram.

@michaelahern
You provided valuable feedback for the DPU, may you have a look to the SHP2 as well?

first message "cmdFunc":12,"cmdId":1,
second message "cmdFunc":12,"cmdId":32,

```

message first_1_pack 
{
	optional float homePower = 1; 
	optional sint32 sysTimezone = 2;
	optional string sysTimezoneId = 4; 
}

message first_2_pack 
{
	optional bytes unknown1 = 1; 
	optional bytes unknown2 = 2; 
}

message first_3_pack 
{
	optional bytes  unknown1 = 1; 
	optional int32 unknown2 = 2; 
	optional first_3_20_pack unknown20 = 20; 
	optional first_3_20_pack unknown21 = 21; 
}

message first_3_20_pack 
{
	optional int32 unknown1 = 1; 
	optional int32 unknown2 = 2; 
}

message first_4_pack 
{
	optional int32 unknown1 = 1; 
	optional string unknown11 = 11; 
	optional int32 unknown21 = 21; 

}

message first_5_pack 
{
	optional int32 unknown1 = 1;
	optional int32 unknown3 = 3;
	optional int32 unknown4 = 4;

}

message first_shp_message
{
	optional first_1_pack  first_1_pack = 1; 
	optional first_2_pack  first_2_pack = 2; 
	optional first_3_pack  first_3_pack = 3; 
	optional first_4_pack  first_4_pack = 4; 
	optional first_5_pack  first_5_pack = 5; 
}

message second_30_5_pack 
{
	optional int32 loadIsEnable1 = 1;
	optional int32 loadPriority2 = 2;
	optional int32 loadPriorityType = 3;
}

message second_30_pack 
{
	optional int32 unknown30_1 = 1; 
	optional int32 unknown30_4 = 4; 
	repeated second_30_5_pack chaSta = 5; 
}


message second_80_1_pack 
{
	repeated bytes string = 1;
}

message acin1_pack 
{
	optional int32 online1 = 1; 
	optional int32 unknown2 = 2; 
	optional int32 unknown3 = 3; 
	optional int32 unknown4 = 4; 
	optional int32 unknown5 = 5; 
	optional int32 unknown6 = 6; 
	optional int32 type = 7; 
}

message name_pack
{
	optional string name = 1;
}

message device_pack 
{
	optional name_pack device = 1;
	optional int32 type2 = 2;
	optional int32 capacity3 = 3;
	optional float ratePower4 = 4;
}

message acin2_pack 
{
	optional device_pack device = 1;
	optional int32 isEnable2 = 2;
	optional int32 isConnect3 = 3;
	optional int32 unknown4 = 4; 
	optional int32 unknown5 = 5; 
	optional int32 unknown6 = 6; 
	optional int32 unknown7 = 7; 
	optional int32 batteryPercentage8 = 8;
	optional float unknown9 = 9; 
	optional int32 unknown10 = 10; 
	optional int32 unknown11 = 11; 
	optional int32 unknown12 = 12; 
	optional int32 unknown13 = 13; 
	optional int32 emsBatTemp14 = 14;
	optional int32 unknown15 = 15; 
	optional int32 unknown16 = 16; 
	optional int32 unknown17 = 17; 
	optional int32 unknown18 = 18; 
	optional int32 unknown19 = 19; 
}

message ac3in_pack 
{
	optional string unknown1 = 1; 
	optional int32 unknown2 = 2; 
	optional int32 unknown3 = 3; 
	optional int32 unknown4 = 4;
	optional float unknown5 = 5;
	optional int32 unknown6 = 6;
	optional float unknown7 = 7; 
	optional int32 unknown8 = 8;
	optional int32 unknown9 = 9;
	optional int32 unknown10 = 10;
	optional int32 unknown11 = 11;
	optional int32 unknown12 = 12;
	optional int32 unknown13 = 13;
}

message second_80_pack 
{
	optional second_80_1_pack unknown_80_1 = 1; 
	optional int32 unknown2 = 2; 
	optional int32 unknown3 = 3; 
	optional int32 unknown4 = 4; 
	optional int32 unknown5 = 5; 
	optional acin1_pack acIn1 = 60; 
	optional acin1_pack acIn2 = 61; 
	optional acin1_pack acIn3 = 62; 
	optional acin2_pack energyInfos1 = 80;
	optional acin2_pack energyInfos2 = 81; 
	optional acin2_pack energyInfos3 = 82; 
	optional ac3in_pack  data1 = 90; 
	optional ac3in_pack  data2 = 91; 
}

message second_81_1_1_pack 
{
	optional int32 id1 = 1;
}

message splitphaseinfo_pack 
{
	optional int32 linkMark1 = 1;
	optional int32 linkCh2 =2;
}

message splitphase_pack 
{
	optional splitphaseinfo_pack splitphaseinfo = 1;
	optional int32 size2 = 2; 
	optional int32 isEnable3 = 3; 
	optional string chName4 = 4;
	optional int32 iconNum5 = 5;
}

message splitphase2_pack 
{
	optional int32 unknown1 = 1; 
	optional int32 unknown2 = 2; 
	optional int32 unknown3 = 3; 
	optional int32 unknown4 = 4; 
	optional int32 unknown5 = 5; 
	optional int32 unknown6 = 6; 
	optional int32 unknown7 = 7; 
	optional int32 unknown8 = 8; 
}

message second_81_1_pack 
{
	optional second_81_1_1_pack strings1 = 1;
	optional string unknown2= 2; 
	optional int32 unknown3 = 3; 
	optional int32 unknown4 = 4; 
	optional splitphase_pack ch1data = 30; 
	optional splitphase_pack ch2data = 31;
	optional splitphase_pack ch3data = 32;
	optional splitphase_pack ch4data = 33;
	optional splitphase_pack ch5data = 34;
	optional splitphase_pack ch6data = 35;
	optional splitphase_pack ch7data = 36;
	optional splitphase_pack ch8data = 37;
	optional splitphase_pack ch9data = 38;
	optional splitphase_pack ch10data = 39;
	optional splitphase_pack ch11data = 40;
	optional splitphase_pack ch12data = 41;

	optional splitphase2_pack ch1data2 = 90;
	optional splitphase2_pack ch2data2 = 91;
	optional splitphase2_pack ch3data2 = 92;
	optional splitphase2_pack ch4data2 = 93;
	optional splitphase2_pack ch5data2 = 94;
	optional splitphase2_pack ch6data2 = 95;
	optional splitphase2_pack ch7data2 = 96;
	optional splitphase2_pack ch8data2 = 97;
	optional splitphase2_pack ch9data2 = 98;
	optional splitphase2_pack ch10data2 = 99;
	optional splitphase2_pack ch11data2 = 100;
	optional splitphase2_pack ch12data2 = 101;
}

message second_81_pack
{
	optional second_81_1_pack unknown_81_1 = 1;
}

message second_82_pack 
{
	optional second_81_1_1_pack errCode1 = 1; 
	optional int32 unknown2 = 2; 
	optional int32 unknown3 = 3; 
	optional int32 unknown4 = 4; 
	optional int32 unknown5 = 5; 
	optional int32 unknown6 = 6; 
	optional int32 unknown7 = 7; 
	optional int32 unknown16 = 16;
	optional float current30 = 30;
	optional float current31 = 31;
	optional int32 unknown32 =32;
	optional float voltage33 = 33;
	optional float voltage34 = 34;
	optional int32 unknown35 =35;
	optional float current36 = 36;
	optional float current37 = 37;
	optional int32 unknown38 =38;
}

message second_shp_message 
{
	optional int32 gridVol = 1;
	optional int32 gridFreq= 2;
	optional int32 unknown3 = 3; 
	optional int32 unexpected4 = 4;
	optional int32 epsMode = 5; 
	optional int32 chargeLimit = 6; 
	optional int32 chargePower = 7; 
	optional int32 unknown8 = 8; 
	optional int32 unknown9 = 9; 
	optional int32 maxCurrent = 10; 
	optional int32 unknown11 = 11; 
	optional int32 maxChargePower = 12; 
	optional int32 unknown14 = 14; 
	optional int32 unknown15 = 15; 
	optional int32 unknown16 = 16; 
	optional string area17 = 17;
	optional int32 unknown18 = 18; 
	optional int32 unknown19 = 19; 
	optional int32 unknown20 = 20; 
	optional int32 unknown21 = 21; 
	optional int32 unixtime22 = 22; 
	optional int32 unknown23 = 23; 
	optional int32 unknown24 = 24; 
	optional int32 unknown25 = 25; 
	optional int32 unknown26 = 26; 
	optional int32 unknown27 = 27; 
	optional int32 unknown28 = 28; 
	optional second_30_pack second_30_pack= 30;
	optional bytes unknown31 = 31; 
	optional bytes unknown32 = 32;
	optional bytes unknown33 = 33;
	optional bytes unknown34 = 34;
	optional bytes unknown35 = 35;
	optional bytes unknown36 = 36;
	optional bytes unknown37 = 37;
	optional bytes unknown38 = 38;
	optional bytes unknown39 = 39;
	optional bytes unknown40 = 40; 
	optional int32 unknown41 = 41; 
	optional int32 unknown61 = 61; 
	optional int32 unknown62 = 62; 
	optional int32 backupReserveSoc63 = 63;
	optional int32 unixtime64 = 64; 
	optional float unknown65 = 65; 
	optional float unknown66 = 66; 
	optional int32 unknown67 = 67; 
	optional int32 unknown70 = 70; 
	optional int32 maxPowerGenerator71 = 71; 
	optional second_80_pack second_80_pack= 80; 
	optional second_81_pack second_81_pack= 81; 
	optional second_82_pack second_82_pack= 82; 
	optional int32 unknown90 = 90; 
	optional int32 unknown91 = 91; 
	optional int32 unknown92 = 92; 
	optional int32 unknown93 = 93; 
	optional int32 unknown94 = 94; 
	optional int32 unknown95 = 95; 
	optional int32 unknown96 = 96; 
	optional int32 unknown97 = 97; 
	optional int32 unknown98 = 98; 
	optional int32 unknown99 = 99; 
	optional int32 unknown100 = 100; 
	optional int32 unknown101 = 101; 
	optional int32 unknown102 = 102; 
	optional int32 unknown103 = 103;
	optional int32 unknown104 = 104;  
}
```

**Comment by Ne0-Hack3r** [comment]
Hey everyone. Jumping in a bit late here but interested in this effort. At this point, does anyone have some python they are willing to share that captures the binary payloads, and formats the data (at least as far as it's been parsed to this point)? I think is could be useful to build a python based binary->JSON adapter that would make it easy to incorporate these encoded payloads into existing projects that are build around parsing JSON and allow for creating YAML based sensors in Home Assistant.

Important note: If you have a developer account at Ecoflow with access to the open API you can get JSON payloads for SHP2 and DPU. Neither are documented or officially supported by Ecoflow (yet) but the data is there. That may help accelerate filling in the blanks on the encoding and/or creating our own community adapter that taps into the real time data stream via the public MQTT.

I'd be happy to setup a repo for this specific effort where we can all contribute code. I have the impression all new products from Ecoflow (Delta Pro 3 for example) will use encoded payload. This makes sense as the BLE connection to these devices and the TCP port open on some older units (like Delta Pro) already use it and it reduces the size of data packets going to/from app/server/device. I suspect once we have a complete "map" of the bytes the comprise each sensor it would be applicable to projects that want to use a local BLE or TCP connection (instead of cloud/MQTT) as well.

It may be possible to derive the maps from a decompiled APK of Android version of the Ecoflow mobile app. I suspect this is how the original hassio-ecoflow project obtained the information to create the local TCP based integration for the gen 1 Delta and River devices. However, the code is highly obfuscated and I've had only limited success with this approach. I am not a java programmer or pro developer and have little experience tracing through decompiled code. Someone with the right background and experience might be able to use this route and possibly provide the mappings to the rest of us.

**Comment by boasist** [comment]
I have the DPU with three batteries (if that matters), SHP2, an android phone, and am awaiting approval of the Ecoflow Developer. I have no coding experience, but can follow a guide and provide any data needed.

**Comment by foxthefox** [comment]
The developer account might be useful to overcome the protobuf challenge and worth trying.
At least the JSON telegrams comming from there might be worth for comparison to the protobuf structure.
So if you have the possibility to capture via the developer account I am interested in th telegrams.

Recently I started to work on power ocean, which also uses protobuf and there was additionally a capture of the webportal in JSON. Thathelps to figuring out the data and giving it the correct names.

At some point I have doubts that all data is available through the developer account. At least for the power stream the documentation is only explaining one message type, but there is another heartbeat message and the messages related to the energy values (reset on daily basis, but very useful for the dashboard).
If we have data we can compare protobuf and JSON.

**Comment by foxthefox** [comment]
@Ne0-Hack3r 
A binary->JSON converter would still need an environment for login and autorization and then interfacing the converter, I guess. The YAML would enable the configuration of the needed values out of the JSON. And maybe there could also some autogeneration of the YAML. Could be feasable.
I would suggest to use the discovery feature of the HA-MQTT integration instead. The datapoints are created on the fly and as an additional benefit the data can be structured in a hierarchy. The hierarchy helps a lot to follow the different telegrams and the nesting of data.
This approach I have implemented as a gateway between my iobroker adapter and the HA instance.
If you want to see it live, have a look to  https://github.com/tolwi/hassio-ecoflow-cloud/issues/193?notification_referrer_id=NT_kwDOAQD7q7M5MDc5NjIwNTI5OjE2ODQxNjQz#issuecomment-2054085389
SHP and DPro is implemented ;-)

**Comment by michaelahern** [comment]
> @michaelahern You provided valuable feedback for the DPU, may you have a look to the SHP2 as well?

@foxthefox Take a look at the attached Protobuf file descriptor and run it through https://protobuf-decoder.netlify.app/. It's not fully decoding for some reason I haven't been able to figure out, but the majority of it decodes and should help you fill in many of the blanks in the message structures you noted above.

[pd303.txt](https://github.com/user-attachments/files/16103352/pd303.txt)

**Comment by foxthefox** [comment]
Thanks a lot! I checked it shortly and there are some left over bytes, maybe the stream is cutted before the regular end. But it contains a majority of values, so I can map/merge it.

**Comment by foxthefox** [comment]
The file descriptor follows the schema and now we have
"ProtoTime" message "cmdFunc":12,"cmdId":1,
"ProtoPushAndSet" message "cmdFunc":12,"cmdId":32,

which I used to decode the telegrams and it looks promising.
@RyanWor 
you may want to have a look into the google sheets, where I have pasted the nicely decoded string beside the 4 outputs of debug4. I hope that the values are plausible.

I you capture another set of telegrams, I am prepared to decode them.

**Comment by RyanWor** [comment]
Hi @foxthefox once again apologies for the delay. We were hit with a hurricane Monday and I have been without power most the week. 

I am looking at your updates in the google sheet now and I can certainly help confirm some of these for you. 

I also setup iobroker with your adapter and this looks like it will be way easier for me to monitor value changes as I modify settings.

**Comment by foxthefox** [comment]
I heard about hurricane and hope that the EF setup did support you.

There is no timeline for finishing something here, it takes the time which is needed.
I appreciate your effort and I am happy that you managed the iobroker installation.

In terms of checking the new output you can put it to a texteditor or so and look at the name and value, if it sounds reasonable. I have no doubt on the unit and correct mapping, but there might be wrong multiplications. But this we can also see in the implementation and correct it quickly.

Yesterday I completed some implementation of the power ocean and today is my plan to implement the first things for SHP2. So now I can feel some pressure ;-) The biggest part is defining the datapoints, therefore I will start with a smaller amount and hope some benefit from SHP work.
I will kepp you updated

**Comment by clayhands** [comment]
Hello, following along to this DPU support thread with interest. 
Is there any room for an inexperienced programmer to help with any data gathering or testing? I have a Delta Pro Ultra as well as basic home assistant setup, which are not currently connected, but if there's anything I can do to help collect or test I am willing.
Or if it's better to just wait until this is all figured out, I will just offer thanks for the work on this.

**Comment by foxthefox** [comment]
Testing and information gathering is always appreciated or when it comes to consistency checks.
I am almost done with defining the datapoints (with first guess of unit and range).
The next will be the correct data handling.

**Comment by rutalreja-deloitte** [comment]
Hi if need be I can test and contribute to the effort, just let me know where I can help, just got my 2xDPUs and 2xESHP2 installed today, would love to get both working in HAOS
My primary goals, control/monitor the ESHP2 circuits, and add the batteries charge consumption to the energy dashboard.

**Comment by foxthefox** [comment]
I am happy to announce the I finished the implementation of SHP2 within the workaround of the iobroker adapter and its gateway to HA.
The version 0.0.35 for iobroker.ecoflow-mqtt is available on git/npm.

Actually there are no commands implemented yet, since they are not traced in the telegrams.

The tests should focus on:
- values and their units (sometimes a multiplication might be needed, or unit is wrong)
- plausibility of value and its description to the real world 
- status interpretation (if its ending with "?" or "..not part of array" then it is a first guess or an unknown value 

@RyanWor as you have iobroker already running, you can update the adapter via the octacat-icon and select/type Ecoflow after click on npm-/git-icon

**Comment by foxthefox** [comment]
In the meantime I have created the version 0.0.37 which solves some issues that the datapoints for discovery were not fully created.

**Comment by rutalreja-deloitte** [comment]
can u help me how do I get 0.0.37 version in HA, is there an easy way to just install via a custom repo in HACS or how? would really like to test it out

The Gist you have above in [comment](https://github.com/tolwi/hassio-ecoflow-cloud/issues/193#issuecomment-1905679243) only tells you how to do it in node red, is there no better way to do this in HAOS?

**Comment by foxthefox** [comment]
Unfortunately it is not that easy. It is neither a HACS nor HA integration. You need to have an instance of ioBroker running (whatever you like as platform, raspberry, Proxmox, Windows...)
see my post from above:
https://github.com/tolwi/hassio-ecoflow-cloud/issues/193#issuecomment-2054085389

I see the need that I am creating a short description

**Comment by DPattee** [comment]
Did any documentation on the Delta Pro Ultra 'workaround' ever get posted?

**Comment by foxthefox** [comment]
Sorry, I did spent my rare time on more functionality of the adapter.
Thanks for reminding me, I will setup a fresh homelab and document accordingly.
Hopefully to the end of the week I will have it together.

**Comment by foxthefox** [comment]
I have made a small picturized description.
I hope it helps

https://github.com/foxthefox/ioBroker.ecoflow-mqtt/blob/main/doc/en/IOB_HA/navi.md

DPU and SHP2 is supported!

**Comment by boasist** [comment]
> I have made a small picturized description. I hope it helps
> 
> https://github.com/foxthefox/ioBroker.ecoflow-mqtt/blob/main/doc/en/IOB_HA/navi.md
> 
> DPU and SHP2 is supported!

Thank you very much, IOB is installed on docker, running, reporting in to HA and connected. The eco flow items are reporting in to HA, yet without any status data, all say unavailable. Running 1.2.2 with a dpu, slave, and shp2 to report it. Ecoflow Mqtt is all filled in and reporting properly.

**Comment by foxthefox** [comment]
First step seems to be accomplished.

I need some more information for helping.

### Most of the checking we have to do is on the io ioBroker instance.
* What status is shown at the adapter instance?

![Image](https://github.com/user-attachments/assets/6635933e-e001-4e0f-a125-ab814df05300)

* if you go to Objects and the ecoflow-mqtt instance (the first is .0), are there values shown?

![Image](https://github.com/user-attachments/assets/cca42adb-0d62-4c69-834e-92c56f3ca67d)

* in the info section should be some "good" status reported and the values for msgPstream should increase in your case

![Image](https://github.com/user-attachments/assets/c6267a44-21ae-465a-8e04-6560bd9e76af)

* Otherwise, what does the log show?

### Checking HA should be on the Settings-> Devices -> MQTT
 
* if HA is connected is shown on IOBconnector:

![Image](https://github.com/user-attachments/assets/d2750e76-d052-4ab6-afe0-ce8499b99b89)

* when the ioBroker adapter is receiving data from EF servers, it reports as connected

![Image](https://github.com/user-attachments/assets/7bef67ba-5145-4c56-b242-c95df08c9672)


I will add a troubleshooting description as well.

**Comment by boasist** [comment]
All is on track with your notes aside from the info on the msg stream - on those and others it says null.

**Comment by foxthefox** [comment]
Its strange that msgCountPstream is null and the status shows it is connected.

I can not judge much what could be the problem, I now need a log.
Please
- put the adapter in debug mode
- enable the logging for DPU only
- enable the msgSetGet and msgUpdate

After save and close the adapter restarts.
I would need the log from the time of restart and the following minute.

**Comment by boasist** [comment]
<!-- Uploading "iobroker.2025-01-26.pdf"... —>`

Here you go.

**Comment by foxthefox** [comment]
Ok, checked the log and there is
"could not subscribe to topics Error: Subscribe error: Unspecified error" suspicious.

So the EF-server is not willing to accept the subscription.

May be there is the solution with changing the mqtts://mqtt-e.ecoflow.com to mqtts://mqtt.ecoflow.com
which was the solution on another issue.
You could verify the server address by https://energychain.github.io/site_ecoflow_mqtt_credentials/ 
it should return the right address

**Comment by boasist** [comment]
Changing that did get more data coming in from the DPU, the SHP though is all unknown. It shows connected, yet all of the data of each is unknown.

I've shared the new log with you as well.

**Comment by foxthefox** [comment]
Thank you.
What I can see:
- there is no longer a subscription problem (so change to other address helped)
- DPU is getting data and it is processed -> problem solved
- less logging on SHP2, but there is something at info level

It would be good to have a next log, where:
- device logging of DPU is disabled 
- device logging of SHP2 is enabled
- same settings for msgSetGet and msgUpdate

**Comment by boasist** [comment]
New log sent as requested.

**Comment by foxthefox** [comment]
Thank you, SHP2 is communicating, thats the good message.
Unfortunately there seems to be a content which brings the adapter to fail when parsing the data.
And also here are some unknown messages comming in, which need some decoding work.

I will make an update and come back when ready.

**Comment by foxthefox** [comment]
I found the parsing problem.

You can try the github version (1.2.2) via expert mode and octacat icon.
Data should come in now for SHP2.

the unknown message of DPU and SHP2 is not included yet.

**Comment by rutalreja-deloitte** [comment]
@foxthefox thank you was able to test 1.2.2 and connected to iobroker and reporting to HAOS. 2 SHP2s and 3DPUs all reporting.
DPU dont constantly update but do when the app is opened.

Naming in HA transfers the Serial Number/ID maybe we can do the name in the device section instead to name the entities
Also next step would be to add the Power IN/OUT from SHP2 to Home Assistant Energy Config. I will try to build a helper but if the MQTTT entities can be aligned to be statistics category maybe it will show up directly. If you have any suggestions please let us know.

**Comment by foxthefox** [comment]
@rutalreja-deloitte 
Thanks for the feedback, nice to hear that it is running. You may want to star my repo.

The updating seems to be a general problem, which some users have.
I think it must be found out what is keeping the sending alive or the app does differently within the communication.

I did not understand your naming problem completely.
I managed to have a structure in mqtt hierarchy:
iobGateway -> device -> subdevice -> entity
All things comming in from EF I flatten to device / module / entity i.e. device = serial number, module = inv/mppt/message name and entity like pv1volt.
This triple set is used to provide a unique ID and should stay.
The device+module is a next hierarchy (subdevice) which gets the serial and the module name.
The entity has then the name of the "signal".

R331ZEB4ZEA12345_bmsMaster_amp = R331ZEB4ZEA12345 serial of Delta 2 example, bmsMaster as module inside Delta2 and amp as the identifier inside the bmsMaster telegram.

As an example.
The device:
{
	"topic": "homeassistant/binary_sensor/R331ZEB4ZEA12345_status/config",
	"payload": {
		"unique_id": "R331ZEB4ZEA12345_main",
		"device": {
			"identifiers": "R331ZEB4ZEA12345_status",
			"manufacturer": "Ecoflow",
			**"name": "R331ZEB4ZEA12345",**
			"model": "Delta 2, 1024Wh",
			"via_device": "iob_0_status"
		},
		"device_class": "connectivity",
		"availability_topic": "iob_ef/iob/info/status",
		"payload_available": "online",
		"payload_not_available": "offline",
		"payload_on": "online",
		"payload_off": "offline",
		"state_topic": "iob_ef/R331ZEB4ZEA12345/info/status",
		"name": "Device Connection to Cloud",
		"origin": { "name": "iobroker.ecoflow-mqtt", "sw_version": "1.2.1" }
	}
}

Each datapoint
{
	"topic": "homeassistant/sensor/R331ZEB4ZEA12345_bmsMaster/amp/config",
	"payload": {
		"has_entity_name": true,
		"unique_id": "R331ZEB4ZEA12345_bmsMaster_amp",
		"state_topic": "iob_ef/R331ZEB4ZEA12345_bmsMaster/amp",
		"device": {
			"identifiers": "R331ZEB4ZEA12345_bmsMaster",
			**"name": "R331ZEB4ZEA12345_bmsMaster",**
			"manufacturer": "Ecoflow",
			"model": "Delta 2, 1024Wh",
			"via_device": "R331ZEB4ZEA12345_status"
		},
		"availability_topic": "iob_ef/R331ZEB4ZEA12345/info/status",
		"payload_available": "online",
		"payload_not_available": "offline",
		"origin": { "name": "iobroker.ecoflow-mqtt", "sw_version": "1.2.1" },
		**"name": "Current",**
		"unit_of_measurement": "A",
		"device_class": "current",
		"state_class": "measurement",
		"value_template": "{{value|round(3)}}"
	}
}

Magically HA creates "R331ZEB4ZEA12345_bmsMaster Current" as a name and leaves out the first name of the device.

Skipping the name (i.e. serial) out of the name makes it really difficult to identify the value.
Can you indicate what I miss?


All values which represent energy can be configured in EnergyFlow card without the need of a helper.
You can go to the configuration of the card and directly all energy sensors are presented in the selection.
It might be advisable to have an intermediate calculation to have a new entity which never changes when your setup is changing, but that is something indiviual and for long term stability of statistics.

**Comment by DPattee** [comment]
I've done some work to set this up today. 

First hurdle was the default mqtt host name needed to be changed. I didn't figure that out until I used https://energychain.github.io/site_ecoflow_mqtt_credentials/ to see that there was no "-e" version in what it returned for me.

Then for the 'naming issue' I agree that using the serial number isn't ideal. In th ioBroke config page, I entered a "Name"  (which matches the friendly name I use in the EcoFlow app) and was expecting the entities to use that rather than the ID. 

I tried changing the name of the entry in the Home Assistant Devices list that was just the ID to be the name, but that only changed one entity name with it. Since there are currently 6 Devices (<id>, and <id>_BPInfo, <id>_action, etc) I need to change those separately I guess. 

On additional thing I noticed is that my Delta Pro Ultra has 2 batteries connected to it, but the "Slave 1" "Slave 2" columns in ioBroker do not have checkboxes under them so I coudln't enable them. 

I see a few different sensors that have battery charge ratings, I'll need to play aroudn to figure out which one reflects which battery as it wasn't immediately obvious.

**Comment by foxthefox** [comment]
Congrats, that you made it.

I think you need to update the adapter to latest version 1.2.2 in order to use the slave1/2 settings fur extra batteries.
You need to enable expert mode and then the custom installation via octacat icon. There you can use non or git. Some instructions with pictures are linked in the Readme in section HA connector.

The naming is right now with the serial number. There is no intention to change this right now. Manually changes in HA might not work, at least it is not tested.

In the Readme I pointed out that there might be a change in the server address needed.

**Comment by ecammit** [comment]
@tolwi , I think this can be closed as resolved also.
